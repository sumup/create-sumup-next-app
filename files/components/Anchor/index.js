import Anchor from './Anchor';

export default Anchor;
