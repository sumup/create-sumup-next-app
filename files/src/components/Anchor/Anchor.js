import React from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { sharedPropTypes } from '@sumup/circuit-ui';

const baseStyles = ({ theme }) => css`
  color: ${theme.colors.p700};
  transition: color 0.1s ease-in-out;

  &:visited {
    color: ${theme.colors.v700};
  }

  &:hover,
  &:focus {
    text-decoration: underline;
    color: ${theme.colors.p500};
  }

  &:active {
    text-decoration: underline;
    color: ${theme.colors.p900};
  }
`;

export const A = styled('a')(baseStyles);

/**
 * A basic anchor component for text links.
 */
function Anchor({ children, title, className, id, ...otherProps }) {
  return (
    <Link {...otherProps}>
      <A {...{ title, className, id }}>{children}</A>
    </Link>
  );
}

Anchor.propTypes = {
  /**
   * The link label.
   */
  children: sharedPropTypes.childrenPropType.isRequired,
  /**
   * Additional information about the link. Displayed as a tooltip.
   * The title attribute is usually ignored by screenreaders.
   */
  title: PropTypes.string,
  /**
   * Pass a class name to customize the style of the anchor.
   */
  className: PropTypes.string,
  /**
   * A unique selector.
   */
  id: PropTypes.string
};

/**
 * @component
 */
export default Anchor;
