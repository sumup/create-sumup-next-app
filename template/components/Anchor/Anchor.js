import React from 'react';
import Link from 'next/link';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { sharedPropTypes } from '@sumup/circuit-ui';

const baseStyles = ({ theme }) => css`
  color: ${theme.colors.p700};
  transition: color 0.1s ease-in-out;

  &:visited {
    color: ${theme.colors.v700};
  }

  &:hover,
  &:focus {
    text-decoration: underline;
    color: ${theme.colors.p500};
  }

  &:active {
    text-decoration: underline;
    color: ${theme.colors.p900};
  }
`;

export const A = styled('a')(baseStyles);

/**
 * A basic anchor component for text links.
 */
function Anchor({ children, title, className, id, ...otherProps }) {
  return (
    <Link {...otherProps} passHref>
      <A {...{ title, className, id }}>{children}</A>
    </Link>
  );
}

/**
 * @component
 */
export default Anchor;
