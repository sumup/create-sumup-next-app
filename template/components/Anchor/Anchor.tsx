import React, { FC } from 'react';
import { css } from '@emotion/core';

import styled from '../../utils/styled';

export type AnchorProps = React.HTMLProps<HTMLAnchorElement>;

const baseStyles = ({ theme }) => css`
  color: ${theme.colors.p700};
  transition: color 0.1s ease-in-out;

  &:visited {
    color: ${theme.colors.v700};
  }

  &:hover,
  &:focus {
    text-decoration: underline;
    color: ${theme.colors.p500};
  }

  &:active {
    text-decoration: underline;
    color: ${theme.colors.p900};
  }
`;

const A = styled.a(baseStyles);

/**
 * A basic anchor component for text links.
 */
export const Anchor: FC<AnchorProps> = (props) => {
  const rel = props.target === '_blank' ? 'noopener noreferrer' : props.rel;
  return <A {...props} rel={rel} />;
};
