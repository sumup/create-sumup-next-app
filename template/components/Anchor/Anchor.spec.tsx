import React from 'react';

import {
  create,
  render,
  renderToHtml,
  axe,
  RenderFn,
} from '../../utils/test-utils';

import { Anchor, AnchorProps } from './Anchor';

describe('Anchor', () => {
  /**
   * Having a separate rendering function for your components makes it easier
   * to render a separate component for each test and reduces boilerplate.
   */
  function renderAnchor(renderFn: RenderFn, props: AnchorProps) {
    return renderFn(<Anchor {...props} />);
  }

  /**
   * Default props help you have sensible defaults that work for most tests.
   * You can pass in custom props to customize rendering for your current test.
   */
  const defaultProps = {
    href: 'https://sumup.com',
    children: `Visit SumUp's website`,
  };

  /**
   * For styled components it can be useful to write snapshot testing.
   * In this case, we want to ensure that an anchor has hover, focus,
   * and active styles.
   */
  it('should render with default styles', () => {
    const actual = renderAnchor(create, defaultProps);
    expect(actual).toMatchSnapshot();
  });

  /**
   * Testing by what your user sees gives you more confidence in your tests.
   * */
  it('should show text inside the anchor', () => {
    const text = `Visit SumUp's website`;
    const { getByText } = renderAnchor(render, {
      ...defaultProps,
      children: text,
    });
    const content = getByText(text, { selector: 'a' });
    expect(content).toBeVisible();
  });

  /**
   * An automatic accessibility test only covers basic best practices.
   * You will still need to test manually to ensure full accessibility.
   * */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderAnchor(renderToHtml, defaultProps);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
