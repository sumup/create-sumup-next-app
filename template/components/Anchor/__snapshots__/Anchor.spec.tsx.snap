// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor should render with default styles 1`] = `
.emotion-0 {
  color: #1760CE;
  -webkit-transition: color 0.1s ease-in-out;
  transition: color 0.1s ease-in-out;
}

.emotion-0:visited {
  color: #8928A2;
}

.emotion-0:hover,
.emotion-0:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #3388FF;
}

.emotion-0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #003C8B;
}

<a
  class="emotion-0"
  href="https://sumup.com"
>
  Visit SumUp's website
</a>
`;
